### ISSUE TITLE
Publishing

### ISSUE THREAD

**Issue by msink** [JetBrains/kotlin-native#2372]
Here: https://github.com/msink/kotlin-libui I'm trying to publish Kotlin/Native library as maven artifact.
For now I managed to publish it to bintray, so it can be used as

```
repositories {
    ...
    maven { url "https://dl.bintray.com/msink/kotlin-native" }
}

kotlin {
    ...
    sourceSets {
        windowsMain {
            kotlin.srcDir('src/nativeMain/kotlin')
            dependencies {
                implementation 'com.github.msink:libui-windows:alpha4.1-2'
            }
        }
        linuxMain {
            kotlin.srcDir('src/nativeMain/kotlin')
            dependencies {
                implementation 'com.github.msink:libui-linux:alpha4.1-2'
            }
        }
        macosxMain {
            kotlin.srcDir('src/nativeMain/kotlin')
            dependencies {
                implementation 'com.github.msink:libui-macosx:alpha4.1-2'
            }
        }
    }
}
```

But still didn't figure out how to publish it as true multiplatform library, for using as
```
    implementation 'com.github.msink:libui:alpha4.1-2'
```

Problem is that it have to be built and publish 3 times - on Linux/Windows/Mac, and every build overwrites `libui` artifact.

How to combine it - I cannot figure out.

PS:
For now uploading works as `./gradlew publishToMavenLocal bintrayUpload`

Maybe `./gradlew publish` supposed to merge, but it fails:
```
> Task :libui:publishKotlinMultiplatformPublicationToMavenRepository
Downloading: com/github/msink/libui/maven-metadata.xml from repository remote at https://bintray.com/msink-native/

> Task :libui:publishKotlinMultiplatformPublicationToMavenRepository FAILED
```

Yes, there isn't `https://bintray.com/msink-native/com/github/msink/libui/maven-metadata.xml` it's in 
`https://dl.bintray.com/msink-native/com/github/msink/libui/maven-metadata.xml`.

**Comment by ilmat192** [comment]
Hi!

The plugin generates a module-file containing variants for all targets declared in a build script (even ones which cannot be built on the current host). See [atomic-fu's moudle-file ](https://kotlin.bintray.com/kotlinx/org/jetbrains/kotlinx/atomicfu/0.11.10-nmpp/atomicfu-0.11.10-nmpp.module) for example.


Looks like in your case the reason why [the module-file](https://dl.bintray.com/msink/kotlin-native/com/github/msink/libui/alpha4.1-2/libui-alpha4.1-2.module) contains only one variant is that you have the following if-chain in your build script:

```
kotlin {
    targets {
        if (os.isWindows()) {
            fromPreset(presets.mingwX64, 'windows') { ... }
        } else if (os.isLinux()) {
            fromPreset(presets.linuxX64, 'linux') { ... }
        } else if (os.isMacOsX()) {
            fromPreset(presets.macosX64, 'macosx') { ... }
        }
    }
}
```

In this case the build script has only one target each time it is executed and the module-file is generated with only one variant. Removing the if-chain should allow publishing the library as a correct multiplatform one:

```
kotlin {
    targets {
        fromPreset(presets.mingwX64, 'windows') { ... }
        fromPreset(presets.linuxX64, 'linux') { ... }
        fromPreset(presets.macosX64, 'macosx') { ... }
    }
}
```

I understand why such an if-chain is added: currently there is no ability to create multiplatform cinterop bindings so we have to write such a hack in build scripts to get an IDE navigation/completion for cinterop bindings.
So it looks like we need a special flag, say `publishMode`, which will enable all targets regardless of the current host. Something like this:

```
kotlin {
    def publishModeEnabled = project.hasProperty('publishMode')
    targets {
        if (os.isWindows() || publishModeEnabled) {
            fromPreset(presets.mingwX64, 'windows') { ... }
        } else if (os.isLinux() || publishModeEnabled) {
            fromPreset(presets.linuxX64, 'linux') { ... }
        } else if (os.isMacOsX() || publishModeEnabled) {
            fromPreset(presets.macosX64, 'macosx') { ... }
        }
    }
}
```

**Comment by msink** [comment]
Thanks.

**Comment by msink** [comment]
Just curious - what is https://dl.bintray.com/msink/kotlin-native/com/github/msink/libui-metadata/ package?
It is not multiplatform, refers only to `libui-macosx`

But seems that it is empty and unused anyway.

**Comment by msink** [comment]
```

**Comment by msink** [comment]
Why it refers to `libui-macosx` package?
(mac build was slowest, and overwrited linux and windows variants)

https://dl.bintray.com/msink/kotlin-native/com/github/msink/libui-metadata/0.1.0/:libui-metadata-0.1.0.module
```
{
  "formatVersion": "0.3",
  "component": {
    "url": "../../libui-macosx/0.1.0/libui-macosx-0.1.0.module",
    "group": "com.github.msink",
    "module": "libui-macosx",
    "version": "0.1.0",
    "attributes": {
      "org.gradle.status": "release"
    }
  },
  "createdBy": {
    "gradle": {
      "version": "4.7",
      "buildId": "n5r3wp7mp5fbpplml44owpp4hm"
    }
  },
  "variants": [
    {
      "name": "metadata-api",
      "attributes": {
        "artifactType": "jar",
        "org.gradle.usage": "kotlin-api",
        "org.jetbrains.kotlin.platform.type": "common"
      },
      "files": [
        {
          "name": "libui-metadata-0.1.0.jar",
          "url": "libui-metadata-0.1.0.jar",
          "size": 261,
          "sha1": "10558791531be1a71c25c4d463ed11336beeac05",
          "md5": "cdb4c31e38e24536e9f84b2038bc0d7f"
        }
      ]
    }
  ]
}
```

**Comment by ilmat192** [comment]
Ah, I see. Just missed it.
It's an interesting side effect of overwriting module-files. It looks like the "root" module should be specified here. I'll take a look, thank you.

**Comment by msink** [comment]
Will it be different in 1.3.20 ?
And what gradle metadata version will be in 1.3.20 - still 0.3 or 0.4 from gradle 4.10+ ?

I tried to build by Kotlin 1.3.20-eap-25 - it works, but .klib is incompatible with 1.3.11, so new publication neded anyway.

**Comment by msink** [comment]
I see no fixes in 1.3.20-eap-25:
- workaround with "if (os.isWindows() || publishModeEnabled) ..." still required for IDEA functionality,
- https://dl.bintray.com/msink/kotlin-dev/com/github/msink/libui-metadata/0.2.0-dev/:libui-metadata-0.2.0-dev.module still refers to "../../libui-macosx/0.2.0-dev/libui-macosx-0.2.0-dev.module"

**Comment by ilmat192** [comment]
Hi! Sorry for the delay with answer.

No, both issues will remain in 1.3.20. The IDE problem is quite complicated and requires some additional design efforts. As for the issue with the links in the metadata files, it doesn't seem to be critical. I didn't dig deep into it so now it's unclear if such a behaviour is a bug/feature of the Gradle metadata system or a side-effect of our machinery built over the top of it.

As for Gralde metadata version, Gradle plugins are not "bound" to any metadata version and most likely `1.3.20` will use the version `0.4`.

**Comment by msink** [comment]
BTW, for me publishing from gradle 5.x works just fine, after this patch:

```diff
     repositories {
         ....
-        maven { url "https://dl.bintray.com/jetbrains/kotlin-native-dependencies" }
+        maven { url "https://dl.bintray.com/touchlabpublic/kotlin" }
     }
     dependencies {
         ...
-        classpath "com.jfrog.bintray.gradle:gradle-bintray-plugin:1.8.2-SNAPSHOT"
+        classpath "com.jfrog.bintray.gradle:gradle-bintray-plugin:1.8.3-kotlin-a1"
     }
 }
```

**Comment by msink** [comment]
I'm lost.
How to translate to Kotlin DSL:
```
    project.publishing.publications.forEach { publication ->
        publication.pom.withXml {
            def root = asNode()
            root.appendNode("name", "libui")
            root.appendNode("description", "Kotlin/Native interop to libui: a portable GUI library")
            root.appendNode("url", POM_SCM_URL)
            root.children().last() + {
                licenses {
                    license {
                        name "MIT License"
                        url POM_SCM_URL
                        distribution "repo"
                    }
                }
                developers {
                    developer {
                        id "msink"
                        name "Mike Sinkovsky"
                        email "msink@permonline.ru"
                    }
                }
                scm {
                    url POM_SCM_URL
                    connection POM_SCM_CONNECTION
                    developerConnection POM_SCM_DEV_CONNECTION
                }
            }
        }
    }
```

**Comment by msink** [comment]
Sure, touchlab version works just fine on Gradle 5.x and Groovy DSL.
But now I'm trying to migrate to Kotlin DSL...

**Comment by ilmat192** [comment]
> I'm lost.
How to translate to Kotlin DSL:

You need to filter only maven publications, since the `pom` property is available only for them. And then you can use the `pom`'s methods. I got something like this:

```
project.publishing.publications.withType(MavenPublication::class.java).forEach { publication ->
    with(publication.pom) {
        withXml {
            val root = asNode()
            root.appendNode("name", "libui")
            root.appendNode("description", "Kotlin/Native interop to libui: a portable GUI library")
            root.appendNode("url", POM_SCM_URL)
        }

        licenses {
            license {
                name.set("MIT License")
                url.set(POM_SCM_URL)
                distribution.set("repo")
            }
        }
        developers {
            developer {
                id.set("msink")
                name.set("Mike Sinkovsky")
                email.set("msink@permonline.ru")
            }
        }
        scm {
            url.set(POM_SCM_URL)
            connection.set(POM_SCM_CONNECTION)
            developerConnection.set(POM_SCM_DEV_CONNECTION)
        }
    }
}
```

**Comment by ilmat192** [comment]
> > most likely 1.3.20 will use the version 0.4

> Is it still true?

It's most probably true but I'm not 100% sure. The library team is working on publication right now.

BTW, thank you and @kpgalligan for the `touchlab` bintary plugin fork. It really helps.

**Comment by msink** [comment]
Thanks, POM works. Now finally bintray-upload:

This works:
```kotlin
bintray {
    user = BINTRAY_USER
    key = System.getenv("BINTRAY_API_KEY")
    override = true // for multi-platform Kotlin/Native publishing
    pkg(delegateClosureOf<BintrayExtension.PackageConfig> {
        userOrg = BINTRAY_USER
        repo = BINTRAY_REPO
        name = "libui"
        setLicenses("MIT")
        vcsUrl = POM_SCM_URL
        version(delegateClosureOf<BintrayExtension.VersionConfig> {
            name = project.version.toString()
            vcsTag = project.version.toString()
            released = Date().toString()
        })
    })
}
```
But my gradle-fu too low for translating this:
```groovy
bintrayUpload.dependsOn publishToMavenLocal

bintrayUpload.doFirst {
    publications = project.publishing.publications.collect {
        println("Uploading artifact '$it.groupId:$it.artifactId:$it.version' from publication '$it.name'")
        it
    }
}
```

**Comment by kpgalligan** [comment]
@ilmat192 Full disclosure, the touchlab fork is a fork of a Jetbrains fork of the bintray plugin. Just needed some minor tweaks for 4.8+. [Here](https://github.com/touchlab/gradle-bintray-plugin) if anybody's interested

**Comment by ilmat192** [comment]
@msink I got something like this (had to cast the publishing task to `BintrayUploadTask` explicitly):

```
val bintrayUpload by tasks.getting
val publishToMavenLocal by tasks.getting

bintrayUpload.dependsOn(publishToMavenLocal)

bintrayUpload.doFirst {
    this as BintrayUploadTask
    this.setPublications(project.publishing.publications.map {
        println("Uploading artifact '$it.groupId:$it.artifactId:$it.version' from publication '$it.name'")
        it
    })
}
```

**Comment by msink** [comment]
Well, this works for me:
```kotlin
val publications = project.publishing.publications.withType(MavenPublication::class.java).map {
    with(it.pom) {
        withXml {
            val root = asNode()
            root.appendNode("name", "libui")
            root.appendNode("description", "Kotlin/Native interop to libui: a portable GUI library")
            root.appendNode("url", POM_SCM_URL)
        }
        licenses {
            license {
                name.set("MIT License")
                url.set(POM_SCM_URL)
                distribution.set("repo")
            }
        }
        developers {
            developer {
                id.set("msink")
                name.set("Mike Sinkovsky")
                email.set("msink@permonline.ru")
            }
        }
        scm {
            url.set(POM_SCM_URL)
            connection.set(POM_SCM_CONNECTION)
            developerConnection.set(POM_SCM_DEV_CONNECTION)
        }
    }
    println("Publishing artifact '${it.groupId}:${it.artifactId}:${it.version}' from publication '${it.name}'")
    it
}

publishing {
    repositories {
        maven { url = uri("https://bintray.com/$BINTRAY_USER/$BINTRAY_REPO") }
    }
}

tasks.withType<BintrayUploadTask>().configureEach {
    dependsOn("publishToMavenLocal")
    setPublications(publications)
}

bintray {
    user = BINTRAY_USER
    key = System.getenv("BINTRAY_API_KEY")
    override = true // for multi-platform Kotlin/Native publishing
    pkg(delegateClosureOf<BintrayExtension.PackageConfig> {
        userOrg = BINTRAY_USER
        repo = BINTRAY_REPO
        name = "libui"
        setLicenses("MIT")
        vcsUrl = POM_SCM_URL
        version(delegateClosureOf<BintrayExtension.VersionConfig> {
            name = project.version.toString()
            vcsTag = project.version.toString()
            released = Date().toString()
        })
    })
}
```

**Comment by msink** [comment]
Well, really it does not work - it runs bintray-upload BEFORE actual build, so only `maven-metadata.xml` uploaded.
Some dependency missed...

**Comment by msink** [comment]
@ilmat192 In your variant `it` have to be cast to `MavenPublication`:
```kotlin
val bintrayUpload by tasks.getting
val publishToMavenLocal by tasks.getting

bintrayUpload.dependsOn(publishToMavenLocal)

bintrayUpload.doFirst {
    this as BintrayUploadTask
    this.setPublications(project.publishing.publications.map {
        it as MavenPublication
        println("Publishing artifact '${it.groupId}:${it.artifactId}:${it.version}' from publication '${it.name}'")
        it
    })
}
```
But it does not work too, same as mine.

So for now reverted that fragment to Groovy.

**Comment by msink** [comment]
With next version based on current master (probably 1.3.30) approach with enabling all targets in publish mode will not work - because now cross-compiling mode is enabled, at least for Linux targets - so now if we want deterministic build we will need to disable Linux target on Windows and Mac CI hosts, and so cannot rely on identical metadata generated on each CI host.
So how publishing will work in upcoming version?

**Comment by ilmat192** [comment]
Currently cross-compilation isn't supported in Gradle plugin. But as for plans according to this feature, the first idea is to provide some setting allowing a user to disable publication of real artifacts for some targets (but still generate correct metadata). We have such mechanisms to handle native targets which cannot be built on current host, but at the moment they are hidden from users. So I think we can expose them to users but I need to double check it.

**Comment by ilmat192** [comment]
Also we can use this [`mavenPublication`](https://kotlinlang.org/docs/reference/building-mpp-with-gradle.html#publishing-a-multiplatform-library) code block.

**Comment by msink** [comment]
@ilmat192 if you mean something like this:
``` Kotlin
tasks.withType<AbstractPublishToMaven>().all {
    onlyIf {
        publication != null && when (publication.name) {
            "windows" -> os.isWindows
            "linux" -> os.isLinux
            "macosx" -> os.isMacOsX
            else -> true
        }
    }
}
```
-- weird, but it does not have any effect - windows host builds and _uploads_ linux target:
https://ci.appveyor.com/project/msink/kotlin-libui/builds/23595921

**Comment by ilmat192** [comment]
It does look weird. Let me take a closer look at it.

**Comment by ilmat192** [comment]
It's because the bintray plugin is used while the example above shows how to disable publishing for `maven-publish`. So the solution is to exclude the undesirable publications from bintray publishing. It can be done when you configure publications of the `bintrayUpload` task. I've managed to do something similar with the following code snippet (Kotlin DSL):

```kotlin
kotlin {
    macosX64("macos")
    linuxX64("linux")
}

bintray {
    // ...
    project.publishing.publications.all {
        if (name != "linux") {
            // Bintray DSL doesn't allow writing just `publication += name` :(
            this@bintray.setPublications(*publications, name)
        }
    }
}
```

If you need to publish the artifacts using `maven-publish` too (e.g. for a local maven repo), you still need to use the approach shown above. There is also a ticket to provide a DSL for disabling some targets in MPP projects: https://youtrack.jetbrains.com/issue/KT-30498.

**Comment by msink** [comment]
So here is publishing [script](https://github.com/msink/kotlin-libui/blob/7266c8d00be3252bfd2d4854c433f262a00afe69/libui/build.gradle.kts):
``` Kotlin
/*
 * Publishing
 */
publishing {
    repositories {
        maven { url = uri("https://bintray.com/${Publish.user}/$BINTRAY_REPO") }
    }
}

val publications = project.publishing.publications.withType(MavenPublication::class.java).map {
    with(it.pom) {
        withXml {
            val root = asNode()
            root.appendNode("name", "libui")
            root.appendNode("description", "Kotlin/Native interop to libui: a portable GUI library")
            root.appendNode("url", Publish.pom.url)
        }
        licenses {
            license {
                name.set("MIT License")
                url.set(Publish.pom.url)
                distribution.set("repo")
            }
            license {
                name.set("The Apache License, Version 2.0")
                url.set("http://www.apache.org/licenses/LICENSE-2.0.txt")
            }
        }
        developers {
            developer {
                id.set("msink")
                name.set("Mike Sinkovsky")
                email.set("msink@permonline.ru")
            }
        }
        scm {
            url.set(Publish.pom.url)
            connection.set(Publish.pom.connection)
            developerConnection.set(Publish.pom.devConnection)
        }
    }
}

bintray {
    user = Publish.user
    key = System.getenv("BINTRAY_API_KEY")
    override = true // for multi-platform Kotlin/Native publishing
    pkg {
        userOrg = Publish.user
        repo = BINTRAY_REPO
        name = "libui"
        setLicenses("MIT", "Apache-2.0")
        vcsUrl = Publish.pom.url
        version {
            name = project.version.toString()
            vcsTag = project.version.toString()
            released = Date().toString()
        }
    }
    project.publishing.publications.all {
        if ((name != "windows" || os.isWindows) &&
            (name != "macosx" || os.isMacOsX) &&
            (name != "linux" || os.isLinux)
        ) {
            // Bintray DSL doesn't allow writing just `publication += name` :(
            this@bintray.setPublications(*publications, name)
        }
    }
}
```

Maybe not very pretty, but seems to work correctly.

**Comment by msink** [comment]
Well, I'm not sure is it a bug or not - but in [Gradle blog](https://blog.gradle.org/gradle-metadata-1.0) I see:
> Gradle Module Metadata 1.0 enables fine-grained dependency resolution for all Gradle users. 
> Starting with Gradle 5.3, if you are a consumer and the library you use has Gradle metadata published, Gradle will automatically consume any Gradle metadata that is published to Maven or Ivy repositories.

Ok, updated libui to Gradle to 5.3.1 and published in metadata 1.0 format.
Upgraded simple [test](https://github.com/msink/hello-libui) too - and it works just fine with `enableFeaturePreview("GRADLE_METADATA")` in `settings.gradle.kts`.

It is __consumer__ - so should work without this string?
But it doesn't - if comment out that shtring - there is error:
```
> Configure project :
Kotlin Multiplatform Projects are an experimental feature.

> Task :compileKotlinLibui FAILED
e: F:\src\hello-libui\src\libuiMain\kotlin\hello.kt: (1, 8): Unresolved reference: libui
e: F:\src\hello-libui\src\libuiMain\kotlin\hello.kt: (3, 14): Unresolved reference: appWindow
e: F:\src\hello-libui\src\libuiMain\kotlin\hello.kt: (8, 5): Unresolved reference: vbox
e: F:\src\hello-libui\src\libuiMain\kotlin\hello.kt: (9, 30): Unresolved reference: TextArea
e: F:\src\hello-libui\src\libuiMain\kotlin\hello.kt: (11, 9): Unresolved reference: button
e: F:\src\hello-libui\src\libuiMain\kotlin\hello.kt: (12, 13): Unresolved reference: action
e: F:\src\hello-libui\src\libuiMain\kotlin\hello.kt: (20, 18): Unresolved reference: textarea
e: F:\src\hello-libui\src\libuiMain\kotlin\hello.kt: (21, 13): Unresolved reference: readonly
e: F:\src\hello-libui\src\libuiMain\kotlin\hello.kt: (22, 13): Unresolved reference: stretchy

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':compileKotlinLibui'.
> Process 'command 'C:\Program Files\Java\jdk1.8.0_111\bin\java.exe'' finished with non-zero exit value 1

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 4s
2 actionable tasks: 2 executed
```

**Comment by ilmat192** [comment]
I'm not sure if it's a bug or not either. Let me investigate it deeper.

**Comment by msink** [comment]
Well, issue with consumer usage without `enableFeaturePreview("GRADLE_METADATA")` seems to be fixed in Kotlin 1.3.50 with Gradle 5.6.

Another question - do I understand correctly that this shuld do exactly same as snippet above, but without ugly tricks with "jfrog.bintray" plugin?
For me it works exactly same.

```kt
publishing {
    publications.withType<MavenPublication>().all {
        pom {
            withXml {
                asNode().apply {
                    appendNode("name", "libui")
                    appendNode("description", "Kotlin/Native interop to libui: a portable GUI library")
                    appendNode("url", Publish.pom.url)
                }
            }
            licenses {
                license {
                    name.set("MIT License")
                    url.set(Publish.pom.url)
                    distribution.set("repo")
                }
                license {
                    name.set("The Apache License, Version 2.0")
                    url.set("http://www.apache.org/licenses/LICENSE-2.0.txt")
                }
            }
            developers {
                developer {
                    id.set("msink")
                    name.set("Mike Sinkovsky")
                    email.set("msink@permonline.ru")
                }
            }
            scm {
                url.set(Publish.pom.url)
                connection.set(Publish.pom.connection)
                developerConnection.set(Publish.pom.devConnection)
            }
        }
    }

    repositories {
        maven {
            url = uri("https://api.bintray.com/maven/${Publish.user}/$BINTRAY_REPO/libui/;publish=0;override=1")
            credentials {
                username = Publish.user
                password = System.getenv("BINTRAY_API_KEY")
            }
        }
    }
}

tasks.withType<AbstractPublishToMaven>().all {
    onlyIf {
        (!name.startsWith("publishWindows") || os.isWindows) &&
        (!name.startsWith("publishMacosx") || os.isMacOsX) &&
        (!name.startsWith("publishLinux") || os.isLinux)
    }
}
```

**Comment by ilmat192** [comment]
Yes, it's possible to publish artifacts to bintray using only the `maven-publish` plugin. You only need to construct the correct URL (just like in your example). The snipped you provided looks correct.
